daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/JsonLedgerClient.scala

```diff
@@ -294,29 +313,41 @@ class JsonLedgerClient(
     )
   }
 
-  // Check that the parties in the token match the given parties.
-  private def validateTokenParties(parties: OneAnd[Set, Ref.Party], what: String): Future[Unit] = {
-    import scalaz.std.string._
-    val tokenParties = Set(tokenPayload.readAs ++ tokenPayload.actAs: _*)
+  // Check that the parties in the token provide read claims for the given parties
+  // and return explicit party specifications if required.
+  private def validateTokenParties(
+      parties: OneAnd[Set, Ref.Party],
+      what: String,
+  ): Future[Option[QueryParties]] = {
+    val tokenParties = tokenPayload.readAs.toSet union tokenPayload.actAs.toSet
+    // Use toSet twice to convert from Party to String
+    val partiesSet = parties.toSet.toSet[String]
+    val missingParties = partiesSet diff tokenParties
     // First check is just for a nicer error message and would be covered by the second
     if (tokenParties.isEmpty) {
       Future.failed(
         new RuntimeException(
           s"Tried to $what as ${parties.toList.mkString(" ")} but token contains no parties."
         )
       )
-    } else if (tokenParties === parties.toSet.toSet[String]) {
-      Future.unit
+    } else if (missingParties.isEmpty) {
+      import scalaz.std.string._
+      if (partiesSet === tokenParties) {
+        Future.successful(None)
+      } else {
+        Future.successful(Some(QueryParties(parties)))
+      }
     } else {
-      Future.failed(new RuntimeException(s"Tried to $what as ${parties.toList
-        .mkString(" ")} but token provides claims for ${tokenParties.mkString(" ")}"))
+      Future.failed(new RuntimeException(s"Tried to $what as [${parties.toList
+        .mkString(", ")}] but token provides claims for [${tokenParties
+        .mkString(", ")}]. Missing claims: [${missingParties.mkString(", ")}]"))
     }
```

Comment:
I would have prefer to have the successful case at the end, not between the two error cases

Reply:
good suggestion 4d79b939cf62044b47835d9f8818ca59893c1f3b